Src/HALAL/Services/Communication/Ethernet/TCP/Socket.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Socket.cpp
+ *
+ *  Created on: Nov 23, 2022
+ *      Author: stefa
+ */
+
+#include "Communication/Ethernet/TCP/Socket.hpp"
+
+unordered_map<EthernetNode,Socket*> Socket::connecting_sockets = {};
+
+Socket::Socket() = default;
+
+Socket::Socket(IPV4 local_ip, uint32_t local_port, IPV4 remote_ip, uint32_t remote_port):remote_ip(remote_ip), remote_port(remote_port){
+	connection_control_block = tcp_new();
+	tcp_bind(connection_control_block, &local_ip.ip_address, local_port);
+	tcp_nagle_disable(connection_control_block);
+	state = INACTIVE;
+	EthernetNode remote_node(remote_ip, remote_port);
+	connecting_sockets[remote_node] = this;
+	tcp_connect(connection_control_block, &remote_ip.ip_address , remote_port, connect_callback);
+}
+
+Socket::Socket(string local_ip, uint32_t local_port, string remote_ip, uint32_t remote_port):Socket(IPV4(local_ip),local_port,IPV4(remote_ip),remote_port){}
+
+
+Socket::Socket(EthernetNode local_node, EthernetNode remote_node):Socket(local_node.ip, local_node.port, remote_node.ip, remote_node.port){}
+
+void Socket::close(){
+  tcp_arg(socket_control_block, nullptr);
+  tcp_sent(socket_control_block, nullptr);
+  tcp_recv(socket_control_block, nullptr);
+  tcp_err(socket_control_block, nullptr);
+  tcp_poll(socket_control_block, nullptr, 0);
+
+  if(tx_packet_buffer!=nullptr){
+	  pbuf_free(tx_packet_buffer);
+  }
+  if(rx_packet_buffer!=nullptr){
+	  pbuf_free(rx_packet_buffer);
+  }
+
+  tcp_close(socket_control_block);
+}
+
+void Socket::reconnect(){
+	EthernetNode remote_node(remote_ip, remote_port);
+	if(!connecting_sockets.contains(remote_node)){
+		connecting_sockets[remote_node] = this;
+	}
+	tcp_connect(connection_control_block, &remote_ip.ip_address , remote_port, connect_callback);
+}
+
+
+void Socket::send(){
+	uint16_t is_freed;
+	pbuf* temporal_packet_buffer;
+	err_t error = ERR_OK;
+	while(error == ERR_OK && tx_packet_buffer != nullptr && tx_packet_buffer->len <= tcp_sndbuf(socket_control_block)){
```

Comment:
```suggestion
	while(error == ERR_OK && tx_packet_buffer != nullptr && tx_packet_buffer->len <= tcp_sndbuf(socket_control_block)){
```

tcp_sndbuf?? No entiendo porque se abrevia esto, si puedes cambiarlo ponle tcp_send_buffer que no cuesta nada y se lee mejor